txt/config.txt
= syslinux.cfg(5) =
:doctype: manpage
:revdate: 2012-10-28
:author: H. Peter Anvin
:author-email: hpa@zytor.com
:editor1: Gene Cumm
:editor1-email: gene.cumm@gmail.com
:editor1-revlast: 2012-10-28
:nbsp8: &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;
:nbsp32: {nbsp8}{nbsp8}{nbsp8}{nbsp8}
:data-uri:

== NAME ==
syslinux.cfg - Syslinux configuration file


//[FIXME:Inspect for '' ** use]


== DESCRIPTION ==
Configuration for the boot behavior and user experience of Syslinux boot loaders, the format of display files and the boot prompt behavior.

Blank lines are ignored.

Note that the configuration file is not completely decoded.  Syntax different from the one described above may still work correctly in this version of Syslinux, but may break in a future one.


== LOCATION/NAME ==
SYSLINUX (before 4.00) used the configuration filename of syslinux.cfg.  EXTLINUX (merged into SYSLINUX as of 4.00) used the filename extlinux.conf.  Both default to searching for the config file in the installed directory (containing ldlinux.sys/extlinux.sys).  As of 4.00, SYSLINUX will search for extlinux.conf then syslinux.cfg in each directory before falling back to the next directory.

As of 3.35, SYSLINUX also searches /boot/syslinux, /syslinux and /.

ISOLINUX (before 4.02) used the configuration filename of isolinux.cfg, searching  /boot/isolinux (starting 2.00), then /isolinux and /.  As of 4.02, ISOLINUX will search for isolinux.cfg then syslinux.cfg in /boot/isolinux before searching for the same files in /isolinux, /boot/syslinux, /syslinux, and /.


== GLOBAL DIRECTIVES - MAIN ==
*#* comment::
A line comment.  As of version 3.10, the space between the *#* and the comment is no longer required.

*MENU* any string::
A directive for the simple menu system, treated as a comment outside the menu.  See menu.txt.

*INCLUDE* 'filename'::
Inserts the contents of another file at this point in the configuration file. Files can currently be nested up to 16 levels deep, but it is not guaranteed that more than 8 levels will be supported in the future.

*DEFAULT* 'kernel' 'options...'::
Sets the default command line.  If Syslinux boots automatically, it will act just as if the entries after DEFAULT had been typed in at the 'boot:' prompt.  Multiple uses will result in an override.
+
If no configuration file is present, or no DEFAULT or *UI* entry is present in the config file, an error message is displayed and the 'boot:' prompt is shown.

*UI* 'module' 'options...'::
Selects a specific user interface module (typically menu.c32 or vesamenu.c32).  The command-line interface treats this as a directive that overrides the DEFAULT and PROMPT directives.  Multiple uses will result in an override.

*LABEL* 'mylabel'::
Begin a new *LABEL* clause.  If 'mylabel' is entered as the kernel to boot, Syslinux should instead boot "image" (specified by a directive from *KERNEL-LIKE DIRECTIVES*) with any specified *DUAL-PURPOSE DIRECTIVES* being used instead of the global instance.
+
'mylabel' should be restricted to non-white-space characters.  Prior to version 3.32, this would transformed to a DOS compatible format of 8.3 with a restricted character set.  A *LABEL* clause must contain exactly 1 of the *KERNEL-LIKE DIRECTIVES* and may contain 1 each of the *LABEL-ONLY DIRECTIVES* or *DUAL-PURPOSE DIRECTIVES*.
+
Within a *LABEL*, using multiple *KERNEL-LIKE DIRECTIVES* or reuse of *LABEL-ONLY DIRECTIVES* or *DUAL-PURPOSE DIRECTIVES* will result in an override.  Otherwise, multiple instances of the same directive will result in the last being effective.


== DUAL-PURPOSE DIRECTIVES ==
Use of any of the *DUAL-PURPOSE DIRECTIVES* as *GLOBAL DIRECTIVES* is discouraged if there will be any non-Linux images loaded as *ALL* images will get these, including those manually entered at the 'boot:' prompt.

*APPEND* 'options...'::
Add one or more options to the kernel command line.  These are added both for automatic and manual boots.  The options are added at the very beginning of the kernel command line, usually permitting explicitly entered kernel options to override them.  This is the equivalent of the LILO "append" option.
+
Use of the parameter 'initrd=' supports multiple filenames separated by commas (ie 'initrd=initrd_file1,initrd_file2') within a single instance. This is mostly useful for initramfs, which can be composed of multiple separate cpio or cpio.gz archives.
+
Note: all initrd files except the last one are zero-padded to a 4K page boundary.  This should not affect initramfs.
+
Note: Only the last effective 'initrd=' parameter is used for loading initrd files.

*APPEND* -::
Append nothing.  APPEND with a single hyphen as argument in a LABEL section can be used to override a global APPEND.

//[FIXME: Shorten subdefinitions]
*IPAPPEND* 'flag_val'::
(PXELINUX only) The IPAPPEND option is available only on PXELINUX.  The flag_val is an OR (sum) of the following integer options:

ifndef::doctype-manpage[[horizontal]]
*1*::: An option of the following format should be generated, based on the input from the DHCP/BOOTP or PXE boot server and added to the kernel command line(see note below):
+
----
ip=<client-ip>:<boot-server-ip>:<gw-ip>:<netmask>
----
+
NOTE:  The use of option 1 is no substitute for running a DHCP client in the booted system and should instead only be used to seed the client.  Without regular renewals, the lease acquired by the PXE BIOS will expire, making the IP address available for reuse by the DHCP server.
+
*2*::: An option of the following format should be generated, in dash-separated hexadecimal with leading hardware type (same as for the configuration file; see pxelinux.txt.) and added to the kernel command line, allowing an initrd program to determine from which interface the system booted:
+
----
BOOTIF=<hardware-address-of-boot-interface>
----
+
*4*::: An option of the following format should be generated, in lower case hexadecimal in the format normally used for UUIDs (same as for the configuration file; see pxelinux.txt.) and added to the kernel command line:
+
----
SYSUUID=<system uuid>
----


== KERNEL-LIKE DIRECTIVES ==
// Alpha sort after KERNEL and LINUX
*KERNEL* 'image'::
Load a kernel-like file 'image' with automatic filetype detection based on file extension, listed under the non-auto-detecting directives, defaulting to *LINUX*.

//[FIXME: Should "'image' as " be removed entirely or added to all?  *LINUX* is used as an example]
*LINUX* 'image'::
Load 'image' as a Linux-like kernel. MEMDISK is an example of a non-Linux kernel loaded in a Linux-like fashion.

*BOOT* 'image'::
(ISOLINUX only: .bin; SYSLINUX only: .bs) Load a boot sector.  .bin is a "CD boot sector" [ISOLINUX only] and .bs is a regular disk boot sector [SYSLINUX only].

*BSS* 'image'::
(SYSLINUX only: .bss) Load a BSS image, a .bs image with the DOS superblock patched in.

*COMBOOT* 'image'::
(.com, .cbt) Load a Syslinux COMBOOT image.  .com images may also be runnable from DOS while .cbt images are not.  See also *comboot.txt*

*COM32* 'image'::
(.c32) Load a Syslinux COM32 (32-bit *COMBOOT*) image.  See also *comboot.txt*

*CONFIG* 'image'::
Load a new configuration file.  The configuration file is read, the working directory is changed (if specified via an APPEND), then the configuration file is parsed.

*FDIMAGE* 'image'::
(Removed as of 4.05, added 1.65; ISOLINUX only: .img) Load a disk image.

*LOCALBOOT* 'type'::
Attempt a different local boot method.  The special value -1 causes the boot loader to report failure to the BIOS, which, on recent BIOSes, should mean that the next boot device in the boot sequence should be activated.  Values other than those documented may produce undesired results.
+
On PXELINUX, "type" 0 means perform a normal boot.  "type" 4 will perform a local boot with the Universal Network Driver Interface (UNDI) driver still resident in memory.  Finally, "type" 5 will perform a local boot with the entire PXE stack, including the UNDI driver, still resident in memory. All other values are undefined.  If you don't know what the UNDI or PXE stacks are, don't worry -- you don't want them, just specify 0.
+
On ISOLINUX, the "type" specifies the local drive number to boot from; 0x00 is the primary floppy drive and 0x80 is the primary hard drive.
//[FIXME: Can SYSLINUX do type as drive number?]

*PXE* 'image'::
(PXELINUX only: .0) Load a PXE NBP (Network Boot Program) image.  The PXE protocol does not provide any means for specifiying or using a command line or initrd.


== LABEL-ONLY DIRECTIVES ==
*INITRD* 'initrd_file'::
Starting with version 3.71, an initrd can be specified in a separate statement (INITRD) instead of as part of the APPEND statement; this functionally appends "initrd=initrd_file" to the kernel command line.  Like 'initrd=', this also supports multiple comma separated file names (see *APPEND*).


== GLOBAL DIRECTIVES - SECONDARY ==
These are global directives that are of lesser importance, often affecting the user experience and not the boot process.

*ALLOWOPTIONS* 'flag_val'::
If 'flag_val' is 0, the user is not allowed to specify any arguments on the kernel command line.  The only options recognized are those specified in an APPEND statement.  The default is 1.

*IMPLICIT* 'flag_val'::
If 'flag_val' is 0, do not load a kernel image unless it has been explicitly named in a *LABEL* statement.  The default is 1.

*TIMEOUT* 'timeout'::
Indicates how long to wait at the 'boot:' prompt until booting automatically, in units of 1/10 s.  The timeout is cancelled as soon as the user types anything on the keyboard, the assumption being that the user will complete the command line already begun.  A timeout of zero (the default) will disable the timeout completely.

*TOTALTIMEOUT* 'timeout'::
Indicates how long to wait until booting automatically, in units of 1/10 s.  This timeout is *not* cancelled by user input, and can thus be used to deal with serial port glitches or "the user walked away" type situations.  A timeout of zero will disable the timeout completely, this is also the default.
+
Both TIMEOUT and TOTALTIMEOUT can be used together, for example:
+
----
# Wait 5 seconds unless the user types something, but
# always boot after 15 minutes.
TIMEOUT 50
TOTALTIMEOUT 9000
----

*ONTIMEOUT* 'kernel options...'::
Sets the command line invoked on a timeout.  Normally this is the same thing as invoked by "DEFAULT".  If this is specified, then "DEFAULT" is used only if the user presses <Enter> to boot.

*ONERROR* 'kernel options...'::
If a kernel image is not found (either due to it not existing, or because IMPLICIT is set), run the specified command.  The faulty command line is appended to the specified options, so if the ONERROR directive reads as:
+
----
ONERROR xyzzy plugh
----
+
and the command line as entered by the user is:
+
----
foo bar baz
----
+
Syslinux will execute the following as if entered by the user:
+
----
xyzzy plugh foo bar baz
----

*SERIAL* 'port [[baudrate] flowcontrol]'::
Enables a serial port to act as the console.  "port" is a number (0 = /dev/ttyS0 = COM1, etc.) or an I/O port address (e.g. 0x3F8); if "baudrate" is omitted, the baud rate defaults to 9600 bps.  The serial parameters are hardcoded to be 8 bits, no parity, 1 stop bit.
+
"flowcontrol" is a combination of the following bits:
+
....
0x001 - Assert DTR
0x002 - Assert RTS
0x008 - Enable interrupts
0x010 - Wait for CTS assertion
0x020 - Wait for DSR assertion
0x040 - Wait for RI assertion
0x080 - Wait for DCD assertion
0x100 - Ignore input unless CTS asserted
0x200 - Ignore input unless DSR asserted
0x400 - Ignore input unless RI asserted
0x800 - Ignore input unless DCD asserted
....
+
All other bits are reserved.
+
Typical values are:
+
....
    0 - No flow control (default)
0x303 - Null modem cable detect
0x013 - RTS/CTS flow control
0x813 - RTS/CTS flow control, modem input
0x023 - DTR/DSR flow control
0x083 - DTR/DCD flow control
....
+
For the SERIAL directive to be guaranteed to work properly, it should be the first directive in the configuration file.
+
NOTE: "port" values from 0 to 3 means the first four serial ports detected by the BIOS.  They may or may not correspond to the legacy port values 0x3F8, 0x2F8, 0x3E8, 0x2E8.
+
Enabling interrupts (setting the 0x008 bit) may give better responsiveness without setting the NOHALT option, but could potentially cause problems with buggy BIOSes.

*NOHALT* 'flag_val'::
If 'flag_val' is 1, don't halt the processor while idle. Halting the processor while idle significantly reduces the power consumption, but can cause poor responsiveness to the serial console, especially when using scripts to drive the serial console, as opposed to human interaction.

*CONSOLE* 'flag_val'::
If 'flag_val' is 0, disable output to the normal video console. If 'flag_val' is 1, enable output to the video console (this is the default.)
+
Some BIOSes try to forward this to the serial console and sometimes make a total mess thereof, so this option lets you disable the video console on these systems.

*FONT* 'filename'::
Load a font in .psf format before displaying any output (except the copyright line, which is output as ldlinux.sys itself is loaded.)  Syslinux only loads the font onto the video card; if the .psf file contains a Unicode table it is ignored.  This only works on EGA and VGA cards; hopefully it should do nothing on others.

*KBDMAP* 'keymap'::
Install a simple keyboard map.  The keyboard remapper used is *very* simplistic (it simply remaps the keycodes received from the BIOS, which means that only the key combinations relevant in the default layout -- usually U.S. English -- can be mapped) but should at least help people with AZERTY keyboard layout and the locations of = and , (two special characters used heavily on the Linux kernel command line.)
+
The included program keytab-lilo.pl from the LILO distribution can be used to create such keymaps.  The file keytab-lilo.txt contains the documentation for this program.

*DISPLAY* 'filename'::
Displays the indicated file on the screen at boot time (before the boot: prompt, if displayed).  Please see the section below on DISPLAY files.
+
NOTE: If the file is missing, this option is simply ignored.

*SAY* 'message'::
Prints the message on the screen.

*PROMPT* 'flag_val'::
If 'flag_val' is 0, display the boot: prompt only if the Shift or Alt key is pressed, or Caps Lock or Scroll lock is set (this is the default).  If 'flag_val' is 1, always display the boot: prompt.

*NOESCAPE* 'flag_val'::
If 'flag_val' is set to 1, ignore the Shift/Alt/Caps Lock/Scroll Lock escapes.  Use this (together with PROMPT 0) to force the default boot alternative.

*NOCOMPLETE* 'flag_val'::
If flag_val is set to 1, the Tab key does not display labels at the boot: prompt.

//   ...etc...
*F1* 'filename'::
*F2* 'filename'::
*F3* 'filename'::
*F4* 'filename'::
*F5* 'filename'::
*F6* 'filename'::
*F7* 'filename'::
*F8* 'filename'::
*F9* 'filename'::
*F10* 'filename'::
*F11* 'filename'::
*F12* 'filename'::

NEWS
Starting with 1.47, changes marked with SYSLINUX, PXELINUX, ISOLINUX
or EXTLINUX apply to that specific program only; other changes apply
to all derivatives.

Changes in 6.00:
	* Add support for booting from EFI.

Changes in 5.00:
	* com32: Switched from the COM32 object format to ELF as it is
	  a much more powerful format that allows undefined symbols to
	  be resolved at runtime and dynamic loading of module
	  dependencies, which means modules now become shared object
	  files instead of statically linked binaries - reducing both
	  disk space and runtime memory consumption.
	* core: Split non-core functionality into ldlinux.c32, which
	  is an ELF module loaded by the core that contains everything
	  the core doesn't require to boot the system, e.g. config
	  parser, command-line interface, etc.
	* Replaced __intcall() calls with direct function calls now
	  that we can resolve undefined symbols at runtime, thanks to
	  the ELF object support. Now that we no longer need to go
	  through the 16-bit interrupt mechanism we can make full use
	  of the 32-bit execution environment. This change required
	  reimplementing lots of the 16-bit assembly code from core/
	  in C.
	* com32: __com32.cs_bounce is gone now we always run in a
	  32-bit environment once we execute ldlinux.c32.
	* ldlinux: A new "PATH" directive was added to the ldlinux.c32
	  config parser that specifies a colon-separated list of
	  directories to search when attempting to load modules.

Changes in 4.06:
	* Support for NTFS, by Paulo Alcantara.
	* EXTLINUX: more robust device detection, allow user to override.
	* kontron_wdt.c32: Add a new module to enable the hardware
	  watchdog of some Kontron boards. It allows enabling the watchdog
	  and then booting a given image.
	* HDT updated, and now can display images regarding some detection
	  steps. Add postexec command to run a particular entry after
	  HDT's execution, add silent option and various fixes.
	* ifcpu.c32: Detect hypervisor presence.
	* lua.c32: Add dhcp support and support for native Syslinux
	  functions syslinux_config(), syslinux_ipappend_strings() and
	  syslinux_reboot().
	* isohybrid: Workaround for various EFI systems.
	* pxechn.c32, a PXE NBP chainloader.  More versatile alternative
	  to pxechain.com and resolves the PXELINUX -> WDS issue with
	  Microsoft Windows Server 2008R2 (Gene Cumm).
	* btrfs: Fix booting off of a subvolume.
	* com32: Add device tree support.
	* SYSLINUX: Fix relative paths for VFAT. The CONFIG and APPEND
	  directives now support entirely relative paths.

Changes in 4.05:
	* HDT updated, and now supports uploading data to a TFTP
	  server.
	* ISOLINUX: remove the .img file support; it has been broken
	  on virtually all systems since the beginning, and has been
	  totally broken since 4.00 at least.  Use MEMDISK instead.
	* chain.c32: Support chaining ReactOS' FreeLdr (Shao Miller)
	* isohybrid: -m option to add support for Mac EFI booting.
        * ifmemdsk.c32: Choose boot option based on presence of
	  MEMDISK.
	* Remove bogus distributed mk-lba-img binary.
	* The Syslinux project has a new, cool logo by Abi
	  "ixxvil" Rasheed (doc/logo/*).

Changes in 4.04:
	* PXELINUX: Fix handling of unqualified DNS names.
	* PXELINUX: Fix timer bug when PXELINUX might be unloaded
	  (Gene Cumm).
	* core/writedec.inc: Fix duplicate declaration and overflow
	  (Gene Cumm).
	* GCC 4.5 fixes.
	* sample directory: Fix Makefile include (Gene Cumm).
	* ver.com: New universal DOS/COMBOOT application to display
	  version information (includes DRMK) (Gene Cumm).
	* rosh.c32: updated; Using getopt() for internal commands to aid
	  parsing options; Fix bugs in ls; add warm reboot and echo
	  (Gene Cumm).
	* com32: fix a file descriptor leak.
	* gfxboot.c32: handle TEXT..ENDTEXT; error out on no LABELs
	  found (Sebastian Herbszt).
	* Fix booting on non-partitioned devices.
	* MBR, isohybrid: Workaround for a BIOS issue on Acer
	  Travelmate and possibly other machines.
        * COM32: Adding ACPI parsing libary
        * HDT: Release 0.4.1 to support ACPI parsing, 
          improved mutli-core/cpu reporting 
	* LUA: Updating to 5.1.4-2
	* SYSLINUX: core/diskstart.inc: Reset DS after checksum in case
	  it isn't 0 (Gene Cumm).
	* win64: Script update for additional mingw compiler names
	  (Gene Cumm).
	* diag: New directory for diagnostic-related tools.  Add a
	  handoff MBR/VBR and geometry display images (Gene Cumm).
	* MEMDISK: use "mem=" parameter to mark available memory above
	  this point as reserved (core already does alignment) (Gene Cumm).
	* MEMDISK: Additional disk probe checks and debug output
	  (Shao Miller, Gene Cumm).
	* gpxe: add gpxelinuxk.0, based off of undionly.kpxe + new
	  script (Gene Cumm).
	* isohybrid: install the isohdpfx*.bin/isohdppx*.bin files to
	  make isohybrid images in one step with GNU xorriso.
	* PXELINUX: disable a hack that would make localboot work on
	  some machines, but break just about as many.  Some machines
	  which worked with "localboot 0" in previous versions may
	  need "localboot -1" in this one.  If you have a machine
	  which requires "localboot -1", a copy of the dmidecode
	  or sysdump output would be appreciated.
	* Include a set of diagnostics by Gene Cumm.
	* Fixes for gcc 4.6 and binutils 2.21.51.
	* chain.c32: Allow "uuid" as a synonym to "guid".
	* Handle directory names starting with .. for vfat and
	  iso9660.
	* New MENU HIDDENKEY command to provide a one-keystroke way to
	  activate a boot option from a hidden menu intro screen.

Changes in 4.03:
	* Don't hang if no configuration file is found.
	* Better support for booting from MBRs which don't pass
	  handover information.
	* EXTLINUX: Try to be smarter about finding the partition
	  offset.
	* chain.c32: support chainloading Dell Real Mode Kernel (Gene
	  Cumm).
	* chain.c32: fix booting in CHS mode.
	* rosh.c32 updated (Gene Cumm).
	* Fix the -s option to the syslinux/extlinux installer (Arwin
	  Vosselman).

com32/chain/options.c
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include "common.h"
#include "chain.h"
#include "utility.h"
#include "options.h"

struct options opt;

static int soi_s2n(char *ptr, unsigned int *seg,
		       unsigned int *off,
		       unsigned int *ip,
		       unsigned int def)
{
    unsigned int segval = 0, offval, ipval, val;
    char *p;

    offval = def;
    ipval = def;

    segval = strtoul(ptr, &p, 0);
    if (p[0] == ':' && p[1] && p[1] != ':')
	offval = strtoul(p+1, &p, 0);
    if (p[0] == ':' && p[1] && p[1] != ':')
	ipval = strtoul(p+1, NULL, 0);

    val = (segval << 4) + offval;

    if (val < ADDRMIN || val > ADDRMAX) {
	error("Invalid seg:off:* address specified..\n");
	goto bail;
    }

    val = (segval << 4) + ipval;

    if (ipval > 0xFFFE || val < ADDRMIN || val > ADDRMAX) {
	error("Invalid seg:*:ip address specified.\n");
	goto bail;
    }

    if (seg)
	*seg = segval;
    if (off)
	*off = offval;
    if (ip)
	*ip  = ipval;

    return 0;
bail:
    return -1;
}

static void usage(void)
{
    unsigned int i;
    static const char key[] = "Press any key...\n";
    static const char *const usage[] = {
"\
Usage:\n\
    chain.c32 [options]\n\
    chain.c32 {fd|hd}<disk#>{,| }[<part#>] [options]\n\
    chain.c32 mbr{:|=}<id>{,| }[<part#>] [options]\n\
    chain.c32 guid{:|=}<guid>{,| }[<part#>] [options]\n\
    chain.c32 label{:|=}<label> [<part#>] [options]\n\
    chain.c32 boot{,| }[<part#>] [options]\n\
    chain.c32 fs [options]\n\
", "\
\nOptions ('no' prefix specifies default value):\n\
    sect[=<s[:o[:i]]>]   Load sector at <s:o>, jump to <s:i>\n\
                         - defaults to 0:0x7C00:0x7C00\n\
                         - ommited o/i values default to 0\n\
    maps                 Map loaded sector into real memory\n\
    nosetbpb             Fix BPB fields in loaded sector\n\
    nofilebpb            Apply 'setbpb' to loaded file\n\
    nosave               Write adjusted sector back to disk\n\
    hand                 Prepare handover area\n\
    nohptr               Force ds:si and ds:bp to point to handover area\n\
    noswap               Swap drive numbers, if bootdisk is not fd0/hd0\n\
    nohide               Disable all hide variations (also the default)\n\
    hide                 Hide primary partitions, unhide selected partition\n\
    hideall              Hide *all* partitions, unhide selected partition\n\
    unhide               Unhide primary partitions\n\
    unhideall            Unhide *all* partitions\n\
    nofixchs             Walk *all* partitions and fix E/MBRs' chs values\n\
    nokeeppxe            Keep the PXE and UNDI stacks in memory (PXELINUX)\n\
    nowarn               Wait for a keypress to continue chainloading\n\
                         - useful to see emited warnings\n\
    nobreak              Actually perform the chainloading\n\
", "\
\nOptions continued ...\n\
    file=<file>          Load and execute <file>\n\
    seg=<s[:o[:i]]>      Load file at <s:o>, jump to <s:i>\n\
                         - defaults to 0:0x7C00:0x7C00\n\
                         - ommited o/i values default to 0\n\
    isolinux=<loader>    Load another version of ISOLINUX\n\
    ntldr=<loader>       Load Windows NTLDR, SETUPLDR.BIN or BOOTMGR\n\
    reactos=<loader>     Load ReactOS's loader\n\
    cmldr=<loader>       Load Recovery Console of Windows NT/2K/XP/2003\n\
    freedos=<loader>     Load FreeDOS KERNEL.SYS\n\
    msdos=<loader>       Load MS-DOS 2.xx - 6.xx IO.SYS\n\
    msdos7=<loader>      Load MS-DOS 7+ IO.SYS\n\
    pcdos=<loader>       Load PC-DOS IBMBIO.COM\n\
    drmk=<loader>        Load DRMK DELLBIO.BIN\n\
    grub=<loader>        Load GRUB Legacy stage2\n\
    grubcfg=<filename>   Set alternative config filename for GRUB Legacy\n\
    grldr=<loader>       Load GRUB4DOS grldr\n\
    bss=<filename>       Emulate syslinux's BSS\n\
    bs=<filename>        Emulate syslinux's BS\n\
\nPlease see doc/chain.txt for the detailed documentation.\n\
"
    };
    for (i = 0; i < sizeof(usage)/sizeof(usage[0]); i++) {
	if (i) {
	    error(key);
	    wait_key();
	}
	error(usage[i]);
    }
}

void opt_set_defs(void)
{
    memset(&opt, 0, sizeof(opt));
    opt.sect = true;	    /* by def. load sector */
    opt.maps = true;	    /* by def. map sector */
    opt.hand = true;	    /* by def. prepare handover */
    opt.brkchain = false;   /* by def. do chainload */
    opt.foff = opt.soff = opt.fip = opt.sip = 0x7C00;
    opt.drivename = "boot";
#ifdef DEBUG
    opt.warn = true;
#endif
}

int opt_parse_args(int argc, char *argv[])
{
    int i;
    unsigned int v;
    char *p;

    for (i = 1; i < argc; i++) {
	if (!strncmp(argv[i], "file=", 5)) {
	    opt.file = argv[i] + 5;
	} else if (!strcmp(argv[i], "nofile")) {
	    opt.file = NULL;
	} else if (!strncmp(argv[i], "seg=", 4)) {
	    if (soi_s2n(argv[i] + 4, &opt.fseg, &opt.foff, &opt.fip, 0))
		goto bail;
	} else if (!strncmp(argv[i], "bss=", 4)) {
	    opt.file = argv[i] + 4;
	    opt.bss = true;
	    opt.maps = false;
	    opt.setbpb = true;
	    /* opt.save = true; */
	} else if (!strncmp(argv[i], "bs=", 3)) {
	    opt.file = argv[i] + 3;
	    opt.sect = false;
	    opt.filebpb = true;
	} else if (!strncmp(argv[i], "isolinux=", 9)) {
	    opt.file = argv[i] + 9;
	    opt.isolinux = true;
	    opt.hand = false;
	    opt.sect = false;
	} else if (!strncmp(argv[i], "ntldr=", 6)) {
	    opt.fseg = 0x2000;  /* NTLDR wants this address */
	    opt.foff = 0;
	    opt.fip = 0;
	    opt.file = argv[i] + 6;
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if (!strncmp(argv[i], "reactos=", 8)) {
	    /*
	     * settings based on commit
	     *   ad4cf1470977f648ee1dd45e97939589ccb0393c
	     * note, conflicts with:
	     *   http://reactos.freedoors.org/Reactos%200.3.13/ReactOS-0.3.13-REL-src/boot/freeldr/notes.txt
	     */
	    opt.fseg = 0;
	    opt.foff = 0x8000;
	    opt.fip = 0x8100;
	    opt.file = argv[i] + 8;
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if (!strncmp(argv[i], "cmldr=", 6)) {
	    opt.fseg = 0x2000;  /* CMLDR wants this address */
	    opt.foff = 0;
	    opt.fip = 0;
	    opt.file = argv[i] + 6;
	    opt.cmldr = true;
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if (!strncmp(argv[i], "freedos=", 8)) {
	    opt.fseg = 0x60;    /* FREEDOS wants this address */
	    opt.foff = 0;
	    opt.fip = 0;
	    opt.sseg = 0x1FE0;
	    opt.file = argv[i] + 8;
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if ( (v = 6, !strncmp(argv[i], "msdos=", v) ||
		     !strncmp(argv[i], "pcdos=", v)) ||
		    (v = 7, !strncmp(argv[i], "msdos7=", v)) ) {
	    opt.fseg = 0x70;    /* MS-DOS 2.00 .. 6.xx wants this address */
	    opt.foff = 0;
	    opt.fip = v == 7 ? 0x200 : 0;  /* MS-DOS 7.0+ wants this ip */
	    opt.sseg = 0x8000;
	    opt.file = argv[i] + v;
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if (!strncmp(argv[i], "drmk=", 5)) {
	    opt.fseg = 0x70;    /* DRMK wants this address */
	    opt.foff = 0;
	    opt.fip = 0;
	    opt.sseg = 0x2000;
	    opt.soff = 0;
	    opt.sip = 0;
	    opt.file = argv[i] + 5;
	    /* opt.drmk = true; */
	    opt.setbpb = true;
	    /* opt.save = true; */
	    opt.hand = false;
	} else if (!strncmp(argv[i], "grub=", 5)) {
	    opt.fseg = 0x800;	/* stage2 wants this address */
	    opt.foff = 0;
	    opt.fip = 0x200;
	    opt.file = argv[i] + 5;
	    opt.grub = true;
	    opt.hand = false;
	    opt.sect = false;
	} else if (!strncmp(argv[i], "grubcfg=", 8)) {
	    opt.grubcfg = argv[i] + 8;
	} else if (!strncmp(argv[i], "grldr=", 6)) {
	    opt.file = argv[i] + 6;
	    opt.grldr = true;
	    opt.hand = false;
	    opt.sect = false;
	} else if (!strcmp(argv[i], "keeppxe")) {
	    opt.keeppxe = 3;
	} else if (!strcmp(argv[i], "nokeeppxe")) {
	    opt.keeppxe = 0;
	} else if (!strcmp(argv[i], "maps")) {
	    opt.maps = true;
	} else if (!strcmp(argv[i], "nomaps")) {
	    opt.maps = false;
	} else if (!strcmp(argv[i], "hand")) {
	    opt.hand = true;
	} else if (!strcmp(argv[i], "nohand")) {
	    opt.hand = false;
	} else if (!strcmp(argv[i], "hptr")) {
	    opt.hptr = true;
	} else if (!strcmp(argv[i], "nohptr")) {
	    opt.hptr = false;
	} else if (!strcmp(argv[i], "swap")) {
	    opt.swap = true;
	} else if (!strcmp(argv[i], "noswap")) {
	    opt.swap = false;
	} else if (!strcmp(argv[i], "nohide")) {
	    opt.hide = 0;
	} else if (!strcmp(argv[i], "hide")) {
	    opt.hide = 1; /* 001b */
	} else if (!strcmp(argv[i], "hideall")) {
	    opt.hide = 2; /* 010b */
	} else if (!strcmp(argv[i], "unhide")) {
	    opt.hide = 5; /* 101b */
	} else if (!strcmp(argv[i], "unhideall")) {
	    opt.hide = 6; /* 110b */
	} else if (!strcmp(argv[i], "setbpb")) {
	    opt.setbpb = true;
	} else if (!strcmp(argv[i], "nosetbpb")) {
	    opt.setbpb = false;
	} else if (!strcmp(argv[i], "filebpb")) {
	    opt.filebpb = true;
	} else if (!strcmp(argv[i], "nofilebpb")) {
	    opt.filebpb = false;
	} else if (!strncmp(argv[i], "sect=", 5) ||
		   !strcmp(argv[i], "sect")) {
	    if (argv[i][4]) {
		if (soi_s2n(argv[i] + 5, &opt.sseg, &opt.soff, &opt.sip, 0))
		    goto bail;
	    }
	    opt.sect = true;
	} else if (!strcmp(argv[i], "nosect")) {
	    opt.sect = false;
	    opt.maps = false;
	} else if (!strcmp(argv[i], "save")) {
	    opt.save = true;
	} else if (!strcmp(argv[i], "nosave")) {
	    opt.save = false;
	} else if (!strcmp(argv[i], "fixchs")) {
	    opt.fixchs = true;
	} else if (!strcmp(argv[i], "nofixchs")) {
	    opt.fixchs = false;
	} else if (!strcmp(argv[i], "warn")) {
	    opt.warn = true;
	} else if (!strcmp(argv[i], "nowarn")) {
	    opt.warn = false;
	} else if (!strcmp(argv[i], "nobreak")) {
	    opt.brkchain = false;
	} else if (!strcmp(argv[i], "break")) {
	    opt.brkchain = true;
	    opt.file = NULL;
	    opt.maps = false;
	    opt.hand = false;
	} else if (((argv[i][0] == 'h' || argv[i][0] == 'f')
		    && argv[i][1] == 'd')
		   || !strncmp(argv[i], "mbr:", 4)
		   || !strncmp(argv[i], "mbr=", 4)
		   || !strncmp(argv[i], "guid:", 5)
		   || !strncmp(argv[i], "guid=", 5)
		   || !strncmp(argv[i], "label:", 6)
		   || !strncmp(argv[i], "label=", 6)
		   || !strcmp(argv[i], "boot")
		   || !strncmp(argv[i], "boot,", 5)
		   || !strcmp(argv[i], "fs")) {
	    opt.drivename = argv[i];
	    if (strncmp(argv[i], "label", 5))
		p = strchr(opt.drivename, ',');
	    else
		p = NULL;
	    if (p) {
		*p = '\0';
		opt.partition = p + 1;
	    } else if (argv[i + 1] && argv[i + 1][0] >= '0'
		    && argv[i + 1][0] <= '9') {
		opt.partition = argv[++i];
	    }
	} else {
	    usage();
	    goto bail;
	}
    }

    if (opt.grubcfg && !opt.grub) {
	error("grubcfg=<filename> must be used together with grub=<loader>.\n");
	goto bail;
    }

#if 0
    if ((!opt.maps || !opt.sect) && !opt.file) {
	error("You have to load something.\n");
	goto bail;
    }
#endif

    if (opt.filebpb && !opt.file) {
	error("Option 'filebpb' requires a file.\n");
	goto bail;
    }

    if (opt.save && !opt.sect) {
	error("Option 'save' requires a sector.\n");
	goto bail;
    }

    if (opt.setbpb && !opt.sect) {
	error("Option 'setbpb' requires a sector.\n");
	goto bail;
    }

    if (opt.maps && !opt.sect) {
	error("Option 'maps' requires a sector.\n");
	goto bail;
    }

    return 0;
bail:
    return -1;
}

/* vim: set ts=8 sts=4 sw=4 noet: */
